shared-libs/transitions/src/transitions/create_user_for_contacts.js

```diff
@@ -0,0 +1,180 @@
+  }
+};
+
+const createNewUser = async ({ roles }, { _id, name, phone, parent }) => {
+  if (!name) {
+    throw new Error(`Replacement contact [${_id}] must have a name.`);
+  }
+  const username = await generateUniqueUsername(name);
+  const user = {
+    username,
+    token_login: true,
+    roles,
+    phone,
+    place: parent._id,
+    contact: _id,
+    fullname: name,
+  };
+  try {
+    return await users.createUser(user, config.get('app_url'));
+  } catch (err) {
+    if (!err.message || typeof err.message === 'string') {
+      throw err;
+    }
+    const message = err.message.message ? err.message.message : err.message;
+    throw new Error(`Error creating new user: ${JSON.stringify(message)}`);
+  }
+};
+
+const replaceUser = async (originalContact, { username, replacementContactId }) => {
+  try {
+    const [newContact, originalUserSettings] = await Promise
+      .all([
+        getNewContact(replacementContactId),
+        users.getUserSettings({ name: username }),
+      ]);
+
+    await createNewUser(originalUserSettings, newContact);
+    await users.resetPassword(originalUserSettings.name);
+    originalContact.user_for_contact.replace[username].status = USER_CREATION_STATUS.COMPLETE;
+  } catch (e) {
+    originalContact.user_for_contact.replace[username].status = USER_CREATION_STATUS.ERROR;
+    return e;
+  }
+};
+
+const getUsersToReplace = (originalContact) => {
+  return Object
+    .entries(originalContact.user_for_contact.replace)
+    .filter(([, { status }]) => status === USER_CREATION_STATUS.READY)
+    .map(([username, { replacement_contact_id }]) => ({ username, replacementContactId: replacement_contact_id }));
+};
+
+/**
+ * Replace a user whose contact has been marked as ready to be replaced with a new user.
+ */
+module.exports = {
+  name: NAME,
+  asynchronousOnly: true,
+  init: () => {
+    const tokenLogin = config.get('token_login');
+    if (!tokenLogin || !tokenLogin.enabled) {
+      throw new Error(
+        `Configuration error. Token login must be enabled to use the create_user_for_contacts transition.`
+      );
+    }
+
+    const appUrl = config.get('app_url');
+    if (!appUrl) {
+      throw new Error(
+        'Configuration error. The app_url must be defined to use the create_user_for_contacts transition.'
+      );
+    }
+  },
+  filter: (doc) => {
+    const contactType = contactTypeUtils.getContactType(config.getAll(), doc);
+    if (
+      !contactType
+      || !contactTypeUtils.isPersonType(contactType)
+      || !doc.user_for_contact
+      || !doc.user_for_contact.replace
+    ) {
+      return false;
+    }
+
+    return !!Object
+      .values(doc.user_for_contact.replace)
+      .find(({ status }) => status === USER_CREATION_STATUS.READY);
+  },
+  onMatch: async change => {
+    const usersToReplace = getUsersToReplace(change.doc);
+    const errors = (await Promise.all(usersToReplace.map(user => replaceUser(change.doc, user))))
+      .filter(error => error);
```

Comment:
I have an alternative implementation to propose (using `Promise.allSettled`).  I personally like it a little better than this because the contents of `errors` does not change over time (not a fan of changing state...).  If it still seems hard to read, though, I am fine with the for-loop.

Reply:
I'm not a fan of wrapping an await in () and chaining things after it, Maybe 2-step it, saving all results and then filtering out errors.

Reply:
Done!